crates/engine/tree/src/tree/payload_processor/mod.rs

```diff
@@ -457,6 +458,58 @@ where
             cleared_sparse_trie.lock().replace(cleared_trie);
         });
     }
+
+    /// Updates the execution cache with the post-execution state from an inserted block.
+    ///
+    /// This is used when blocks are inserted directly (e.g., locally built blocks by sequencers)
+    /// to ensure the cache remains warm for subsequent block execution.
+    ///
+    /// The cache enables subsequent blocks to reuse account, storage, and bytecode data without
+    /// hitting the database, maintaining performance consistency.
+    pub(crate) fn on_inserted_executed_block(
+        &self,
+        block_with_parent: BlockWithParent,
+        bundle_state: &BundleState,
+    ) {
+        debug!(target: "engine::caching", ?block_with_parent, "Updating execution cache for inserted block");
+
+        self.execution_cache.update_with_guard(|cached| {
+            // Take existing cache (if any) or create fresh caches
+            let (caches, cache_metrics) = match cached.take() {
+                Some(existing) => {
+                    // Check if existing cache matches parent - if not, restore and skip
+                    if existing.executed_block_hash() != block_with_parent.parent {
+                        warn!(
+                            target: "engine::caching",
+                            cache_hash = %existing.executed_block_hash(),
+                            parent_hash = %block_with_parent.parent,
+                            "Cache parent mismatch for inserted block, skipping cache update"
+                        );
+                        // Restore the cache since we're not updating it
+                        *cached = Some(existing);
+                        return;
+                    }
+                    existing.split()
+                }
+                None => (
+                    ExecutionCacheBuilder::default().build_caches(self.cross_block_cache_size),
+                    CachedStateMetrics::zeroed(),
+                ),
```

Comment:
I see, this is the case we we have no cache yet and this was received before any regular payload.